### DISCUSSION TITLE
Docker Implementing rate limits for pulling Docker images from the Docker Hub Registry

### DISCUSSION THREAD

**Discussion by skreddy6673** [concourse/concourse#6039] [General]
Is this something as a concourse Admin i should be worried about?
More info can be found here https://www.docker.com/blog/scaling-docker-to-serve-millions-more-developers-network-egress/

Please remove this discussion if this is not valid/appropriate.

**Comment by vito** [comment]
Thanks for bringing this up - this has been on the radar, but probably deserves more focused attention.

tl;dr: yes.

Concourse's 1 minute default polling interval will pretty quickly exhaust the 100-200 request limit, especially with multiple resources pointing to Docker Hub. Even though it only does a lightweight `GET` request to detect if the version has changed, and doesn't actually download the image, this will still count as a pull.

The `registry-image` resource type will retry with exponential backoff when it hits this limit (see [here](https://github.com/concourse/registry-image-resource/blob/34f2db54b63ca7ac8c516997c89e37da28af8bed/backoff.go#L14)), but because this limit is per 6 hours and the `registry-image` resource gives up after an hour, you will see failed checks in your pipelines. If the problem ended here, it wouldn't be so bad; you'd have failing checks, but jobs would still be able to run with whatever versions are able to be collected. There would probably "just" be a 6-hour lag time for versions collected from Docker Hub.

The big problem is `image_resource`, which involves a synchronous check call when the task runs. With exponential backoff, this will add a significant delay (and probably an error) before the task even runs. Worse, this `check` call runs with no cursor version (the current version to check from), so we can't even change it to just no-op on 429 and return the cursor version.

I'm actually working on a big refactor that should help with this (following #6022), so that's lucky timing. I guess I'll need to finish this before November 1st. 🤔 With this refactor complete, we could modify `registry-image` to just no-op and return the cursor version if it hits a `429 Too Many Requests` error. This won't help if the rate limit is exceeded before any versions can be found though.

I would also recommend enabling [global resources](https://concourse-ci.org/global-resources.html) if you haven't already - this will reduce checking for equivalent resource definitions across the cluster. This won't help with `image_resource`-defined images though - at least not until I finish my refactor.

Until there's a better solution, the nuclear option is to increase the default checking interval from 1 minute to some very high value, and set a lower [`check_every`](https://concourse-ci.org/resources.html#schema.resource.check_every) interval for resources which need more frequent checking. (Or vice versa - keep the 1 minute default, and set a higher `check_every` for anything pointing to Docker Hub.) Again, this will help for pipeline-defined resources, but won't help with `image_resource`s until the refactor I mentioned is complete.

Taking a step back, I think the intent of these rate limits is to signify an end to the ["free lunch"](https://en.wikipedia.org/wiki/There_ain%27t_no_such_thing_as_a_free_lunch) that Docker Hub appeared to provide for so long. Just 100 pulls per 6 hours is a super restrictive limit for a CI server, especially once with multiple tenants. I don't think this is something you can ignore forever, and this is why they are being very vocal with this announcement and have given folks so much time to adapt.

I would recommend setting up a [registry mirror with a pull-through cache](https://docs.docker.com/registry/recipes/mirror/). It's possible to configure a mirror in the `registry-image` resource via [`registry_mirror`](https://github.com/concourse/registry-image-resource/#source-configuration). Unfortunately it is painful to configure this across all pipelines. There is an open discussion around making this easier: https://github.com/concourse/concourse/pull/5942

Another alternative is to throw money at the problem, and get a Pro plan so that you don't get rate limited at all. The problem is this would require configuring the account's credentials on all your resources, so we would still need a change similar to the one proposed in https://github.com/concourse/concourse/pull/5942#issuecomment-669283962

In addition to that you may want to consider using a different registry for your own images. [GitHub just launched one](https://github.blog/2020-09-01-introducing-github-container-registry/), but I haven't tried it yet. Many IaaSes also provide a container registry. This won't help for commonly used Docker Library images like `golang` and `ubuntu` though.

Hope this helps! We've definitely got a few things to sort out on our end in preparation for this. I would anticipate doing a Concourse upgrade by November 1st in order to have those changes.

**Comment by vito** [comment]
Tracking the work on this in https://github.com/concourse/concourse/issues/6073 + https://github.com/concourse/concourse/projects/119

**Comment by jmccann** [comment]
> I would recommend setting up a registry mirror with a pull-through cache. It's possible to configure a mirror in the registry-image resource via registry_mirror. Unfortunately it is painful to configure this across all pipelines. There is an open discussion around making this easier: #5942

From what I can tell this isn't possible in Concourse 5.x and you'd have to [upgrade to 6.5.1+ to get this](https://github.com/concourse/registry-image-resource/issues/242#issuecomment-705049146)?

Are there any solutions to being able to use a mirror with 5.x?

Also, it seems there [isn't a way to do docker builds with registry mirrors](https://github.com/vito/oci-build-task/issues/37) yet?

**Reply by skreddy6673** [reply]
@jmccann You can use registry_mirror in 5.x, at least i used in v5.7.1. You can use docker-image with registry_mirror like

```
jobs:
- name: job-hello-world
  plan:
  - config:
      container_limits: {}
      image_resource:
        source:
          repository: busybox
          registry_mirror: MIRROR_URL
        type: docker-image
      platform: linux
      run:
        args:
        - hello world
        path: echo
    task: hello-world
  public: true
```

**Reply by skreddy6673** [reply]
https://github.com/concourse/docker-image-resource#source-configuration look for `registry_mirror`.

**Reply by jmccann** [reply]
@skreddy6673 Just to verify, I would need to change `registry-image` and `oci-build-task` both to `docker-image` to pull and/or build images with a resource that supports `registry_mirror` to support this in 5.x?

If so at least that may give me a path forward.  Was hoping there was something simpler I was missing to do it in 5.x.

**Reply by skreddy6673** [reply]
`registry-image` will not support building, so you should be using something like the oci-build task or docker-image.

If you want to use `registry-image` in 5.x for image pull/push you can use it like below

```
jobs:
- name: job-hello-world
  plan:
  - config:
      container_limits: {}
      image_resource:
        source:
          repository: REGISTRY_MIRROR_URL/library/busybox
        type: registry-image
      platform: linux
      run:
        args:
        - hello world
        path: echo
    task: hello-world
  public: true
``` 

The only difference i see in defining `docker-image` and `registry-image` in 5.x is the place where you add the Registry mirror url.

**Comment by jmccann** [comment]
So now that I've updated most references to use a registry mirror per the above another thought crossed my mind.  Is there a way to set registry mirror for the "core" resources in 5.x?  For "custom" `resource_types` I was able to configure it.  Do I need to define "core" `resource_types` like they are "custom" ones in all my pipelines?

e.g.

```diff
+resource_types:
+  - name: git
+    type: registry-image
+    source:
+      repository: docker.mirror.com/concourse/git-resource
+      tag: 1.9.1

resources:
  - name: repo
    type: git
    icon: github-circle
    source:
      uri: https://github.com/concourse/concourse.git
```

And if so, how would I do the same for `registry-image`?

**Reply by skreddy6673** [reply]
This might be the dumb question, what do you mean by "core" resources?

**Reply by jamieklassen** [reply]
@skreddy6673 i believe @jmccann is talking about the "bundled" resource types that Linux-platform concourse workers support by default:

https://github.com/concourse/ci/blob/3199605b7e79ea25710a43c4e12682f8993f88f2/pipelines/reconfigure.yml#L146-L161

this sounds a fair bit like the idea implemented in #6079, which of course doesn't help in 5.x. At least in docker/k8s, the core resource types live under `/usr/local/concourse/resource-types`, where they have a metadata JSON and a tarball of the image. If you wanted to change the default behaviour for a base resource, it's finicky but you could modify that image and replace it on each worker.

**Reply by jmccann** [reply]
@jamieklassen thanks for clarifying the terminology 😄 

Yes, I'm trying to make sure if those are getting pulled that it's pulled through a registry mirroring.  But you are saying they are bundled on the worker.  Does that mean such bundled resources like `git` and `registry-image` etc which are also docker images **don't get pulled**? ... they are already provided and directly on the worker?  If so then I don't have to worry about them.  😄

**Reply by jamieklassen** [reply]
Sorry I joined this conversation late and missed some details. you're right that bundled resource types don't get pulled!

**Reply by mvdkleijn** [reply]
@jamieklassen Is there a way for administrators to add resources to the local bundled set? In other words, could I take a commonly used resource from Docker hub, download it and add it to the worker's local bundle somehow?

**Reply by mvdkleijn** [reply]
> @jamieklassen Is there a way for administrators to add resources to the local bundled set? In other words, could I take a commonly used resource from Docker hub, download it and add it to the worker's local bundle somehow?

Never mind... already answered here I see. :+1: https://github.com/concourse/concourse/discussions/6267
